FreeType/Format/Multiple.hsc

```diff
@@ -0,0 +1,235 @@
+     array is 'null'.
+
+     Please refer to the
+     [Format-Specific API > Multiple Masters](https://www.freetype.org/freetype2/docs/reference/ft2-multiple_masters.html)
+     chapter of the reference.
+
+     Internal: "FreeType.Format.Multiple.Internal".
+ -}
+
+module FreeType.Format.Multiple
+  ( -- ** FT_MM_Axis
+    FT_MM_Axis (..)
+    -- ** FT_Multi_Master
+  , FT_Multi_Master (..)
+    -- ** FT_Var_Axis
+  , FT_Var_Axis (..)
+    -- ** FT_Var_Named_Style
+  , FT_Var_Named_Style (..)
+    -- ** FT_MM_Var
+  , FT_MM_Var (..)
+    -- ** FT_Get_Multi_Master
+  , ft_Get_Multi_Master
+    -- ** FT_Get_MM_Var
+  , ft_Get_MM_Var
+    -- ** FT_Done_MM_Var
+  , ft_Done_MM_Var
+    -- ** FT_Set_MM_Design_Coordinates
+  , ft_Set_MM_Design_Coordinates
+    -- ** FT_Set_Var_Design_Coordinates
+  , ft_Set_Var_Design_Coordinates
+    -- ** FT_Get_Var_Design_Coordinates
+  , ft_Get_Var_Design_Coordinates
+    -- ** FT_Set_MM_Blend_Coordinates
+  , ft_Set_MM_Blend_Coordinates
+    -- ** FT_Get_MM_Blend_Coordinates
+  , ft_Get_MM_Blend_Coordinates
+    -- ** FT_Set_Var_Blend_Coordinates
+  , ft_Set_Var_Blend_Coordinates
+    -- ** FT_Get_Var_Blend_Coordinates
+  , ft_Get_Var_Blend_Coordinates
+    -- ** FT_Set_MM_WeightVector
+  , ft_Set_MM_WeightVector
+    -- ** FT_Get_MM_WeightVector
+  , ft_Get_MM_WeightVector
+    -- ** FT_VAR_AXIS_FLAG_XXX
+  , pattern FT_VAR_AXIS_FLAG_HIDDEN
+    -- ** FT_Get_Var_Axis_Flags
+  , ft_Get_Var_Axis_Flags
+    -- ** FT_Set_Named_Instance
+  , ft_Set_Named_Instance
+  ) where
+
+import           FreeType.Core.Base.Types
+import           FreeType.Core.Types.Types
+import           FreeType.Error.Values
+import           FreeType.Exception.Internal
+import           FreeType.Format.Multiple.Internal
+import           FreeType.Format.Multiple.Types
+
+import           Foreign.Marshal.Array
+import           Foreign.Marshal.Utils
+import           Foreign.Ptr
+import           Foreign.Storable
+import           Language.Haskell.TH.Syntax (Name)
+
+#include "ft2build.h"
+#include FT_MULTIPLE_MASTERS_H
+
+ft_Get_Multi_Master
+  :: FT_Face            -- ^ face
+  -> IO FT_Multi_Master -- ^ master
+ft_Get_Multi_Master =
+  autoAllocaError 'ft_Get_Multi_Master ft_Get_Multi_Master'
+
+
+ft_Get_MM_Var
+  :: FT_Face            -- ^ face
+  -> IO (Ptr FT_MM_Var) -- ^ master
+ft_Get_MM_Var =
+  autoAllocaError 'ft_Get_MM_Var ft_Get_MM_Var'
+
+
+
+ft_Done_MM_Var
+  :: FT_Library    -- ^ library
+  -> Ptr FT_MM_Var -- ^ master
+  -> IO ()
+ft_Done_MM_Var =
+  autoError 'ft_Done_MM_Var ft_Done_MM_Var'
+
+
+
+setCoords
+  :: Storable a
+  => Name
+  -> (FT_Face -> FT_UInt -> Ptr a -> IO FT_Error)
+  -> (FT_Face -> [a] -> IO ())
+setCoords name f face []     =
+  ftError name $ f face 0 nullPtr
+
+setCoords name f face coords =
+  withArray coords $ ftError name . f face (fromIntegral $ length coords)
+
+
+
+getCoords
+  :: Storable a
+  => Name
+  -> (FT_Face -> FT_UInt -> Ptr a -> IO FT_Error)
+  -> (FT_Face -> FT_UInt -> IO [a])
```

Comment:
It's a shame that we're going from unboxed `Ptr a` to boxed `[a]`. This seems like a decision the caller should be allowed to make. What do you think about keeping this interface as `Ptr a`: some callers might want to use `allocaArray` as you have, but others would want (e.g.) the `StorableArray` interface?

Reply:
I do not think this is an issue, as if anyone requires tight loops for efficiency can just rewrite the functions by using definitions in `FreeType.Format.Multiple.Internal`.

Reply:
It's weird to me to export `Internal` modules -- seems like a contradiction, no?

Reply:
They're mostly there to not pile stuff together (`Internal` modules storing foreign imports for functions I wrapped and deprecations), the user is supposed to use `FreeType` and _if they wish_ `FreeType.Internal`. Well and there's also `FreeType.Exception` for `FtError`.